File: app/src/main/java/com/qweex/quite/Options.java
package com.qweex.quite;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.graphics.Color;
import android.net.Uri;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;
import java.util.Arrays;
import java.util.Comparator;

public class Options {
    private AlertDialog popupDialog, filetypesSelector;

    boolean reverseSort;
    TextView aboutText, filetypesText;


    public static Comparator<File> sortOrder;
    public static boolean recurse;
    public static String[] filetypes = new String[
            ImageFragment.filesHandled.length +
                    GifFragment.filesHandled.length +
                    VideoFragment.filesHandled.length];
    public static String[] filetypesSelected = new String[filetypes.length];

    public void updateFiletypesText() {
        int c = 0;
        for(String s : filetypesSelected)
            c += s.equals(".") ? 0 : 1;
        filetypesText.setText("Select filetypes (" + Integer.toString(c) + ")");
    }


    public Options(final MainActivity act) {

        LinearLayout ll = new LinearLayout(act);
        LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        ll.setOrientation(LinearLayout.VERTICAL);
        LinearLayout.LayoutParams lpM = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        lpM.setMargins(40,40,40,40);

        aboutText = new TextView(act);
        CheckBox reverseCheckbox = new CheckBox(act),
                 recurseCheckbox = new CheckBox(act);
        Spinner sortOrderSpinner = new Spinner(act);
        View filetypesItem = act.getLayoutInflater().inflate(android.R.layout.simple_list_item_1, null);
        filetypesText = ((TextView)filetypesItem.findViewById(android.R.id.text1));
        ll.addView(aboutText, lp);
        ll.addView(sortOrderSpinner, lp);
        ll.addView(reverseCheckbox, lpM);
        ll.addView(recurseCheckbox, lpM);
        ll.addView(filetypesItem, lp);

        aboutText.setPadding(40, 40, 40, 40);
        aboutText.setBackgroundColor(Color.parseColor("#FDFDF0"));

        String[] sorts = new String[] {"Order by Name", "Order by Date", "Order by Size", "Order by Type", "Order Random"};
        ArrayAdapter<String> adap = new ArrayAdapter<String>(act, android.R.layout.simple_list_item_1, sorts);
        adap.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        sortOrderSpinner.setAdapter(adap);
        int sel = 0;
        if(sortOrder==dateDesc|| sortOrder==dateAsc)
            sel = 1;
        else if(sortOrder==sizeDesc|| sortOrder==sizeAsc)
            sel = 2;
        else if(sortOrder==typeDesc|| sortOrder==typeAsc)
            sel = 3;
        else if(sortOrder==RANDOM)
            sel = 4;
        sortOrderSpinner.setSelection(sel);
        sortOrderSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                Comparator<File> oldSortOrder = sortOrder;
                switch(position) {
                    case 0:
                        sortOrder = reverseSort ? nameDesc : nameAsc;
                        return;
                    case 1:
                        sortOrder = reverseSort ? dateDesc : dateAsc;
                        return;
                    case 2:
                        sortOrder = reverseSort ? sizeDesc : sizeAsc;
                        return;
                    case 3:
                        sortOrder = reverseSort ? typeDesc : typeAsc;
                        return;
                    case 4:
                        sortOrder = RANDOM;
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });

        reverseCheckbox.setChecked(reverseSort);
        reverseCheckbox.setText("Reverse Order");
        reverseCheckbox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                reverseSort = isChecked;
            }
        });

        recurseCheckbox.setChecked(recurse);
        recurseCheckbox.setText("Recurse");

        //fileTypes
        updateFiletypesText();
        boolean[] checkedFiletypes = new boolean[filetypesSelected.length];
        for(int i=0; i<checkedFiletypes.length; i++)
            checkedFiletypes[i] = !filetypesSelected[i].equals(".");
        AlertDialog.Builder filetypesBuilder = new AlertDialog.Builder(act);
        filetypesBuilder.setMultiChoiceItems(filetypes, checkedFiletypes, new DialogInterface.OnMultiChoiceClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which, boolean isChecked) {
                filetypesSelected[which] = isChecked ? filetypes[which] : ".";
            }
        });
        filetypesBuilder.setTitle("Select Filetypes");
        filetypesBuilder.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                updateFiletypesText();
            }
        });
        filetypesSelector = filetypesBuilder.create();
        filetypesItem.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View v) {
                 Log.d("filetypesText", "showSelector");
                 filetypesSelector.show();
             }
         });

        ScrollView sv = new ScrollView(act);
        sv.addView(ll, lp);

        AlertDialog.Builder builder = new AlertDialog.Builder(act);
        builder.setView(sv);
        builder.setPositiveButton("Apply", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                Uri currentUri = Uri.parse(
                        "file:" + act.qAdapter.files[
                                act.pager.getCurrentItem()
                                ].getPath());
                Log.d("Options", "re-initializing adapter : " + currentUri.getPath());

                //act.getIntent().setData(currentUri);
                //act.initializeView();
                //act.initializeAdapter();

                act.qAdapter.rescan();
                /*
                if(quiteAdapter.fragments!=null) {
                    FragmentTransaction ft = quiteAdapter.fragmentManager.beginTransaction();
                    for (FragmentBase fb : quiteAdapter.fragments)
                        ft.remove(fb);
                    ft.commit();
                }//*/
                act.qAdapter.fragments = new FragmentBase[0];
                act.qAdapter.notifyDataSetChanged();
                act.pager.setAdapter(null);
                act.pager.destroyDrawingCache();
                act.qAdapter.rescan();
                act.qAdapter.notifyDataSetChanged();

                act.qAdapter = new QuiteAdapter(currentUri, act.getSupportFragmentManager(), act);
                if(act.qAdapter.getCount()==0) {
                    Toast.makeText(act, "No images in directory", Toast.LENGTH_SHORT).show();
                    act.finish();
                    return;
                }
                act.pager.setAdapter(act.qAdapter);
                 //*/
                        /*
                Log.d("New Current", quiteAdapter.files[
                        act.pager.getCurrentItem()
                        ].getPath() + "!");
                act.pager.setAdapter(quiteAdapter);
                act.pager.setCurrentItem(
                        act.lastSelected = quiteAdapter.getIndexOfStart()
                );*/

                popupDialog.hide();
               // */
            }
        });
        popupDialog = builder.create();

        // Start slideshow w/ seconds | Stop slideshow

    }

    public void show(FragmentBase fragment) {
        aboutText.setText(fragment.getAbout());
        popupDialog.show();
    }

    public static Comparator<File> nameAsc = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return f1.getPath().compareTo(f2.getPath());
        }
    },
    nameDesc = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return f2.getPath().compareTo(f1.getPath());
        }
    },
    dateAsc = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return (int) (f1.lastModified() - f2.lastModified());
        }
    },
    dateDesc = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return (int) (f2.lastModified() - f1.lastModified());
        }
    },
    sizeAsc = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return (int) (f1.getTotalSpace() - f2.getTotalSpace());
        }
    },
    sizeDesc = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return (int) (f2.getTotalSpace() - f1.getTotalSpace());
        }
    },
    typeAsc = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return f1.getName().substring(f1.getName().lastIndexOf(".")+1).compareTo(
                   f2.getName().substring(f2.getName().lastIndexOf(".")+1));
        }
    },
    typeDesc = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return f2.getName().substring(f2.getName().lastIndexOf(".")+1).compareTo(
                    f1.getName().substring(f1.getName().lastIndexOf(".")+1));
        }
    },
    RANDOM = new Comparator<File>() {
        public int compare(File f1, File f2) {
            return 0;
        }
    };

    static {
        sortOrder = nameAsc;
        int i=0;
        for(String s : ImageFragment.filesHandled)
            filetypes[i++] = s;
        for(String s : GifFragment.filesHandled)
            filetypes[i++] = s;
        for(String s : VideoFragment.filesHandled)
            filetypes[i++] = s;
        Arrays.sort(filetypes);
        for(int j=0; j<filetypes.length; j++)
            filetypesSelected[j] = filetypes[j];
    }

}

File: app/src/main/java/com/qweex/quite/Options.kt
package com.qweex.quite

import android.app.AlertDialog
import android.content.DialogInterface
import android.graphics.Color
import android.net.Uri
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.*
import java.io.File
import java.util.*


class Options(act:MainActivity) {
    private var popupDialog: AlertDialog? = null
    private val filetypesSelector:AlertDialog

    internal var reverseSort:Boolean = false
    internal var aboutText:TextView
    internal var filetypesText:TextView

    fun updateFiletypesText() {
        var c = 0
        for (s in filetypesSelected)
            c += if (s == ".") 0 else 1
        filetypesText.text = "Select filetypes (" + Integer.toString(c) + ")"
    }

    init{

        val ll = LinearLayout(act)
        val lp = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        ll.orientation = LinearLayout.VERTICAL
        val lpM = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        lpM.setMargins(40, 40, 40, 40)

        aboutText = TextView(act)
        val reverseCheckbox = CheckBox(act)
        val recurseCheckbox = CheckBox(act)
        val sortOrderSpinner = Spinner(act)
        val filetypesItem = act.layoutInflater.inflate(android.R.layout.simple_list_item_1, null)
        filetypesText = (filetypesItem.findViewById(android.R.id.text1) as TextView)
        ll.addView(aboutText, lp)
        ll.addView(sortOrderSpinner, lp)
        ll.addView(reverseCheckbox, lpM)
        ll.addView(recurseCheckbox, lpM)
        ll.addView(filetypesItem, lp)

        aboutText.setPadding(40, 40, 40, 40)
        aboutText.setBackgroundColor(Color.parseColor("#FDFDF0"))

        val sorts = arrayOf<String>("Order by Name", "Order by Date", "Order by Size", "Order by Type", "Order Random")
        val adap = ArrayAdapter<String>(act, android.R.layout.simple_list_item_1, sorts)
        adap.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        sortOrderSpinner.adapter = adap
        var sel = 0
        if (sortOrder === dateDesc || sortOrder === dateAsc)
            sel = 1
        else if (sortOrder === sizeDesc || sortOrder === sizeAsc)
            sel = 2
        else if (sortOrder === typeDesc || sortOrder === typeAsc)
            sel = 3
        else if (sortOrder === RANDOM)
            sel = 4
        sortOrderSpinner.setSelection(sel)
        sortOrderSpinner.onItemSelectedListener = object:AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent:AdapterView<*>, view:View, position:Int, id:Long) {
                val oldSortOrder = sortOrder
                when (position) {
                    0 -> {
                        sortOrder = if (reverseSort) nameDesc else nameAsc
                        return
                    }
                    1 -> {
                        sortOrder = if (reverseSort) dateDesc else dateAsc
                        return
                    }
                    2 -> {
                        sortOrder = if (reverseSort) sizeDesc else sizeAsc
                        return
                    }
                    3 -> {
                        sortOrder = if (reverseSort) typeDesc else typeAsc
                        return
                    }
                    4 -> sortOrder = RANDOM
                }
            }

            override fun onNothingSelected(parent:AdapterView<*>) {}
        }

        reverseCheckbox.isChecked = reverseSort
        reverseCheckbox.text = "Reverse Order"
        reverseCheckbox.setOnCheckedChangeListener { buttonView, isChecked -> reverseSort = isChecked }

        recurseCheckbox.isChecked = recurse
        recurseCheckbox.text = "Recurse"

        //fileTypes
        updateFiletypesText()
        val checkedFiletypes = BooleanArray(filetypesSelected.size)
        for (i in checkedFiletypes.indices)
        checkedFiletypes[i] = filetypesSelected[i] != "."
        val filetypesBuilder = AlertDialog.Builder(act)
        filetypesBuilder.setMultiChoiceItems(filetypes, checkedFiletypes, DialogInterface.OnMultiChoiceClickListener { dialog, which, isChecked -> filetypesSelected[which] = if (isChecked) filetypes[which] else "." })
        filetypesBuilder.setTitle("Select Filetypes")
        filetypesBuilder.setPositiveButton("Ok") { dialog, which -> updateFiletypesText() }
        filetypesSelector = filetypesBuilder.create()
        filetypesItem.setOnClickListener {
            Log.d("filetypesText", "showSelector")
            filetypesSelector.show()
        }

        val sv = ScrollView(act)
    sv.addView(ll, lp)

    val builder = AlertDialog.Builder(act)
    builder.setView(sv)
    builder.setPositiveButton("Apply", object:DialogInterface.OnClickListener {
        override fun onClick(dialog:DialogInterface, which:Int) {
            val currentUri = Uri.parse(
            ("file:" + act.qAdapter.files!![act.pager!!.getCurrentItem()].getPath()))
            Log.d("Options", "re-initializing adapter : " + currentUri.path!!)

         //act.getIntent().setData(currentUri);
         //act.initializeView();
         //act.initializeAdapter();

            recurse = recurseCheckbox.isChecked

            act.qAdapter.rescan()

//        if(quiteAdapter.fragments!=null) {
//            FragmentTransaction ft = quiteAdapter.fragmentManager.beginTransaction();
//            for (FragmentBase fb : quiteAdapter.fragments)
//                ft.remove(fb);
//            ft.commit();
//        }
//        act.getQAdapter().fragments = arrayOfNulls<FragmentBase>(0)
//        act.getQAdapter().notifyDataSetChanged()
//        act.getPager().setAdapter(null)
//        act.getPager().destroyDrawingCache()
//        act.getQAdapter().rescan()
//        act.getQAdapter().notifyDataSetChanged()
//
//        act.setQAdapter(QuiteAdapter(currentUri, act.getSupportFragmentManager(), act))
//        if (act.getQAdapter().getCount() === 0)
//        {
//        Toast.makeText(act, "No images in directory", Toast.LENGTH_SHORT).show()
//        act.finish()
//        return
//        }
//        act.getPager().setAdapter(act.getQAdapter())
         //*/
                /*
        Log.d("New Current", quiteAdapter.files[
                act.pager.getCurrentItem()
                ].getPath() + "!");
        act.pager.setAdapter(quiteAdapter);
        act.pager.setCurrentItem(
                act.lastSelected = quiteAdapter.getIndexOfStart()
        );
        */

        popupDialog!!.hide()
        // */
            }
    })
    popupDialog = builder.create()

    // Start slideshow w/ seconds | Stop slideshow

    }

     fun show(fragment:FragmentBase) {
         aboutText.text = fragment.about
         popupDialog!!.show()
    }

    companion object {
         var sortOrder:Comparator<File?>
         var recurse:Boolean = false
         var filetypes = arrayOfNulls<String>(
                (ImageFragment.filesHandled.size +
                GifFragment.filesHandled.size +
                VideoFragment.filesHandled.size))
        var filetypesSelected:Array<String?>

         var nameAsc:Comparator<File?> = Comparator { f1, f2 -> f1!!.path.compareTo(f2!!.path) }
         var nameDesc:Comparator<File?> = Comparator { f1, f2 -> f2!!.path.compareTo(f1!!.path) }
         var dateAsc:Comparator<File?> = Comparator { f1, f2 -> (f1!!.lastModified() - f2!!.lastModified()) as Int }
         var dateDesc:Comparator<File?> = Comparator { f1, f2 -> (f2!!.lastModified() - f1!!.lastModified()) as Int }
         var sizeAsc:Comparator<File?> = Comparator { f1, f2 -> (f1!!.getTotalSpace() - f2!!.getTotalSpace()) as Int }
         var sizeDesc:Comparator<File?> = Comparator { f1, f2 -> (f2!!.getTotalSpace() - f1!!.getTotalSpace()) as Int }
         var typeAsc:Comparator<File?> = Comparator { f1, f2 ->
             f1!!.name.substring(f1.name.lastIndexOf(".") + 1).compareTo(
                     f2!!.name.substring(f2.name.lastIndexOf(".") + 1))
         }
         var typeDesc:Comparator<File?> = Comparator { f1, f2 ->
             f2!!.name.substring(f2.name.lastIndexOf(".") + 1).compareTo(
                     f1!!.name.substring(f1.name.lastIndexOf(".") + 1))
         }
         var RANDOM:Comparator<File?> = Comparator { _, _ -> 0 }

        init{
            filetypesSelected = arrayOfNulls(filetypes.size)
            sortOrder = nameAsc
            var i = 0
            for (s in ImageFragment.filesHandled)
                filetypes[i++] = s
            for (s in GifFragment.filesHandled)
                filetypes[i++] = s
            for (s in VideoFragment.filesHandled)
                filetypes[i++] = s
            Arrays.sort(filetypes)
            for (j in filetypes.indices)
                filetypesSelected[j] = filetypes[j]
        }
    }

}
